core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -371,6 +371,13 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
       throw new SparkException("An application name must be set in your configuration")
     }
 
+    // Thread name has been set to "Driver" if user code ran by AM on a YARN cluster
```

Comment:
Checking in `SparkConf` would be an overhead I think, because it seems `set` would need to check for it. Initializing a `SparkContext` with an invalid `SparkConf` (with `yarn-cluster` setup) would be the best place to deal with it. As you said, checking for system property `spark.yarn.app.id` would be most feasible.